File: java/Play.java
public class Play {
  private static final int WORLD_WIDTH = 150;
  private static final int WORLD_HEIGHT = 40;

  public static void main(String[] args) throws Exception {
    run();
  }

  public static void run() throws Exception {
    var world = new World(
      WORLD_WIDTH,
      WORLD_HEIGHT
    );

    var minimal = System.getenv("MINIMAL") != null;

    if (!minimal) {
      System.out.println(world.render());
    }

    var totalTick = 0.0;
    var lowestTick = Double.MAX_VALUE;
    var totalRender = 0.0;
    var lowestRender = Double.MAX_VALUE;

    while (true) {
      var tickStart = System.nanoTime();
      world.doTick();
      var tickFinish = System.nanoTime();
      var tickTime = (tickFinish - tickStart);
      totalTick += tickTime;
      lowestTick = Math.min(lowestTick, tickTime);
      var avgTick = (totalTick / world.tick);

      var renderStart = System.nanoTime();
      var rendered = world.render();
      var renderFinish = System.nanoTime();
      var renderTime = (renderFinish - renderStart);
      totalRender += renderTime;
      lowestRender = Math.min(lowestRender, renderTime);
      var avgRender = (totalRender / world.tick);

      if (!minimal) {
        System.out.print("\u001b[H\u001b[2J");
      }

      System.out.println(
        "#%d - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)".formatted(
          world.tick,
          _f(lowestTick),
          _f(avgTick),
          _f(lowestRender),
          _f(avgRender)
        )
      );

      if (!minimal) {
        System.out.print(rendered);
      }
    }
  }

  private static double _f(double value) {
    // nanoseconds -> milliseconds
    return value / 1_000_000;
  }
}

File: kotlin/play.kt
public class Play {
  companion object {
    private const val WORLD_WIDTH: Int = 150
    private const val WORLD_HEIGHT: Int = 40

    public fun run() {
      val world = World(
        width = WORLD_WIDTH,
        height = WORLD_HEIGHT,
      )

      val minimal = System.getenv("MINIMAL") != null

      if (!minimal) {
        println(world.render())
      }

      var totalTick = 0.0
      var lowestTick = Double.MAX_VALUE
      var totalRender = 0.0
      var lowestRender = Double.MAX_VALUE

      while (true) {
        val tickStart = System.nanoTime()
        world.doTick()
        val tickFinish = System.nanoTime()
        val tickTime = (tickFinish - tickStart).toDouble()
        totalTick += tickTime
        lowestTick = minOf(lowestTick, tickTime)
        val avgTick = (totalTick / world.tick)

        val renderStart = System.nanoTime()
        val rendered = world.render()
        val renderFinish = System.nanoTime()
        val renderTime = (renderFinish - renderStart).toDouble()
        totalRender += renderTime
        lowestRender = minOf(lowestRender, renderTime)
        val avgRender = (totalRender / world.tick)

        if (!minimal) {
          print("\u001b[H\u001b[2J")
        }

        println(
          String.format(
            "#%d - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)",
            world.tick,
            _f(lowestTick),
            _f(avgTick),
            _f(lowestRender),
            _f(avgRender)
          )
        )

        if (!minimal) {
          print(rendered)
        }
      }
    }

    private fun _f(value: Double): Double {
      // nanoseconds -> milliseconds
      return value / 1_000_000
    }
  }
}

fun main() {
  Play.run()
}
